config.org
#+TITLE: My Emacs Config :)
#+AUTHOR: Luca Muscat

Hi there! As you might have guessed, this is my EMACS config which I use pretty often. There are some basic plugins in here which in my opinion everyone should be using. Hope that you could learn something by going through my config.
* Installation
** Quick Installation
*** Windows
Copy the following script into your command line. Please make sure that you have git installed from before hand.
#+BEGIN_SRC shell
cd %appdata%
git clone https://github.com/lucamuscat/Emacs-Config.git
rename Emacs-Config .emacs.d
cd .emacs.d
rename "./.emacs" "../.emacs"
#+END_SRC

Once you paste this into your cmd, do =M-x eval-expression= and run the following
#+BEGIN_SRC emacs-lisp
(require 'package)
(setq package-enable-at-startup nil)

;;; remove SC if you are not using sunrise commander and org if you like outdated packages
(setq package-archives '(("ELPA"  . "http://tromey.com/elpa/")
			 ("gnu"   . "http://elpa.gnu.org/packages/")
			 ("melpa" . "https://melpa.org/packages/")
			 ("elpa"  . "https://elpa.gnu.org/packages/")
			 ))
(package-initialize)
#+END_SRC

Then call =M-x package-install use-package=. After that, go into the .emacs.d file and run =eval-buffer=.

* Initialization
** GC Thresholds
Increasing =gc-cons-threshold= to a very high number to will decrease the load and compile time. This will later be lowered back down to a more sane value to prevent complications. ([[https://github.com/jwiegley/dot-emacs/blob/master/init.el][taken from J. Wiegley's init.el]])

#+BEGIN_SRC emacs-lisp
(defconst emacs-start-time (current-time))

(defvar file-name-handler-alist-old file-name-handler-alist)

(setq package-enable-at-startup nil
      file-name-handler-alist nil
      message-log-max 16384
      gc-cons-threshold 402653184
      gc-cons-percentage 0.6
      auto-window-vscroll nil)

(add-hook 'after-init-hook
          `(lambda ()
             (setq file-name-handler-alist file-name-handler-alist-old
                   gc-cons-threshold 800000
                   gc-cons-percentage 0.1)
             (garbage-collect)) t)
#+END_SRC

** Preventing the scratch buffer from triggering prog mode
By default, the scratch buffer has a =prog-mode= major mode, this means that any package loaded in with the =prog-mode-hook= will be called unnecessarily. Making the prog-mode-hook counter intuitive to begin with.

#+BEGIN_SRC emacs-lisp
(setq initial-major-mode (quote fundamental-mode))
#+END_SRC

** Benchmark
#+BEGIN_SRC emacs-lisp
(use-package benchmark-init
  :ensure t
  :init(benchmark-init/activate))
#+END_SRC
** Delete Selection when typing
#+BEGIN_SRC emacs-lisp
(delete-selection-mode 1)
#+END_SRC
** Startup optimizations
Taken from [[https://github.com/MatthewZMD/.emacs.d#orgd87dfce][MatthewZMD]]
*** Disable =package-enable-at-startup=
#+BEGIN_SRC emacs-lisp
(setq package-enable-at-startup nil)
#+END_SRC
*** Disable =site-run-file=
#+BEGIN_SRC emacs-lisp
(setq site-run-file nil)
#+END_SRC
* Theme
#+BEGIN_SRC emacs-lisp

(use-package doom-themes
	:diminish
	:ensure t
	:config (load-theme 'doom-city-lights t)
)

#+END_SRC
* Interface Settings
These are built-in UI enhancements.
** Looks
*** Removing startup screen
The start up screen will be replaced with the dashboard package.
#+BEGIN_SRC emacs-lisp
(setq inhibit-splash-screen t
initial-buffer-choice  nil
)
#+END_SRC

*** Display Line Numbers

#+BEGIN_SRC emacs-lisp
(add-hook 'prog-mode-hook 'display-line-numbers-mode)
#+END_SRC

*** Disable menus and scroll bars
Replace the -1 with a 1 keep these features.
#+BEGIN_SRC emacs-lisp
(menu-bar-mode -1)
(toggle-scroll-bar -1)
(tool-bar-mode -1)
#+END_SRC

*** Disable bell
I find this feature annoying, feel free to remove this line if you like the sound of computerized bells
#+BEGIN_SRC emacs-lisp
(setq ring-bell-function 'ignore)
#+END_SRC

*** Clock
A clock which shows in the mode line.
#+BEGIN_SRC emacs-lisp
(setq display-time-24hr-format t)
(setq display-time-format "%d %B %Y - %H:%M")
(display-time-mode 1)
#+END_SRC
*** Dashboard
#+BEGIN_SRC emacs-lisp
(use-package dashboard
:ensure t
:config(setq dashboard-startup-banner "~/.emacs.d/images/KEC_Dark_BK.png")
(setq dashboard-banner-logo-title "Don't do the rain dance if you can't handle the thunder - Ken M")
(setq dashboard-items '((recents  . 3)
                        (bookmarks . 3)
                        (projects . 5)
                        (agenda . 0)
                        (registers . 0)))
(dashboard-setup-startup-hook)
)
#+END_SRC
*** Beacon
Helps user find the line which they are editing when they swap buffer.
#+BEGIN_SRC emacs-lisp
(use-package beacon
	:ensure t
	:config
		(beacon-mode 1)
)
#+END_SRC

*** Customize frame name
 #+BEGIN_SRC emacs-lisp
 (setq-default
 frame-title-format '("Lucinda?"))
 #+END_SRC

*** Customize font-family and font size
#+BEGIN_SRC emacs-lisp
(set-frame-font "Consolas 14" nil t)
#+END_SRC
*** Treemacs
#+BEGIN_SRC emacs-lisp
(use-package treemacs
	:ensure t
	:diminish
	:bind*("C-<tab>" . treemacs)
	:custom
	(treemacs-indentation 2)
	(treemacs-eldoc-display t)
	(treemacs-show-hidden-files t)
	(treemacs-width 35)
	(treemacs-display-in-side-window nil)
	(treemacs-pre-file-insert-predicates)
)

(use-package treemacs-projectile
	:ensure t
	:diminish
	:after treemacs projectile
)

(use-package lsp-treemacs
	:ensure t
	:diminish
	:after treemacs lsp
	:config(lsp-treemacs-sync-mode 1)
)
#+END_SRC
* Functionality
** Awesome pair
https://github.com/manateelazycat/awesome-pair
#+BEGIN_SRC emacs-lisp
(use-package awesome-pair
	:load-path("~/.emacs.d/elpa/")
	:hook( (org-mode prog-mode) . awesome-pair-mode)
	:bind
	("(" . awesome-pair-open-round)
	("[" . awesome-pair-open-bracket)
	("{" . awesome-pair-open-curly)
	("=" . awesome-pair-equal)
	("M-\"" . awesome-pair-wrap-double-quote)
	("M-[" . awesome-pair-wrap-bracket)
	("M-{" . awesome-pair-wrap-curly)
	("M-(" . awesome-pair-wrap-round)
	("M-)" . awesome-pair-unwrap)
)
#+END_SRC
** Projectile
#+BEGIN_SRC emacs-lisp
(use-package projectile
	:diminish
	:ensure t
	:init(projectile-mode)
	(require 'cc-mode)
	:bind("C-x r p" . projectile-switch-project)
		 ("C-x C-M-f" . projectile-find-file)
		 ("C-x C-M-s" . projectile-save-project-buffers)
	:bind(:map c-mode-base-map
		("<f1>" . projectile-compile-project))
		("<f2>" . luca/c-debug)
)

(use-package helm-projectile
	:ensure t
	:after projectile
	:bind("C-M-s" . helm-projectile-ag)
	:init(helm-projectile-on)
)
#+END_SRC
** Magit
#+BEGIN_SRC emacs-lisp
(use-package magit
	:ensure t
	:diminish
	:commands(magit-stage-file magit-status magit-commit-create)
	:bind(:map prog-mode-map
		("C-c t" . magit-stage-file)
		("C-c s" . magit-status)
		("C-c c" . magit-commit-create)
	)
)

(use-package vc
	:no-require t
	:defer t
	:commands(vc diff)
	:bind("C-c d" . vc-diff)
)

(use-package git-gutter
	:ensure t
	:diminish
	:hook (prog-mode . git-gutter-mode)
	:hook (magit-post-refresh . git-gutter:update-all-windows)
	:bind(:map prog-mode-map
		("C-c n" . git-gutter:next-hunk)
 		("C-c p" . git-gutter:previous-hunk)
)
)
#+END_SRC
** UTF-8
This is very important as sometimes you might encounter some different kinds of characters in the the package list which will start throwing annoying errors.
#+BEGIN_SRC emacs-lisp
(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
;; backwards compatibility as default-buffer-file-coding-system
;; is deprecated in 23.2.
(if (boundp 'buffer-file-coding-system)
	(setq-default buffer-file-coding-system 'utf-8)
	(setq default-buffer-file-coding-system 'utf-8))

;; Treat clipboard input as UTF-8 string first; compound text next, etc.
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))
#+END_SRC
** Backups
Although backups are recommended, I don't like how they're implemented (it litters any directory you work in).
#+BEGIN_SRC emacs-lisp
(setq make-backup-files nil)
(setq auto-save-default nil)
#+END_SRC
** Change yes/no prompts to y/n
Normally when the user is prompted to give a yes or no answer, they will have to type out the "yes" or "no" in its entirety. This option changed it to just a "y" for yes and a "n" no (I mean we're using EMACS, might as well save all the key strokes you can ;) )
#+BEGIN_SRC emacs-lisp
(defalias 'yes-or-no-p 'y-or-n-p)
#+END_SRC

** Line wrapping
#+BEGIN_SRC emacs-lisp
(global-visual-line-mode)
#+END_SRC
** Which-key
This is a god send for beginners and novices alike. In short, this allows the user to see what key bindings they can use on the fly.
#+BEGIN_SRC emacs-lisp
(use-package which-key
	:ensure t
	:diminish
	:config (which-key-mode)
	)
#+END_SRC

** ISpell Setup

#+BEGIN_SRC emacs-lisp
(use-package ispell
	:no-require t
	:defer t
	:bind (:map org-mode-map("C-<return>" . ispell-word))
)

#+END_SRC

** Dictionary
#+BEGIN_SRC emacs-lisp
(use-package define-word
	:diminish
	:ensure t
	:defer t
)
#+END_SRC
** helm-ag
=helm-ag= is used to search all of the files in a project.
#+BEGIN_SRC emacs-lisp
(use-package helm-ag
	:ensure t
	:diminish
	:commands(helm-ag)
)
#+END_SRC
** Unbind certain keys
#+BEGIN_SRC emacs-lisp
(global-unset-key "\C-z")
(global-unset-key "\C-x\C-z")
(global-unset-key "\C-x\C-c")
#+END_SRC
** Fill-region
#+BEGIN_SRC emacs-lisp
(use-package fill
	:no-require t
	:bind("<f9>" . fill-region )
)
#+END_SRC
** Undo-tree
#+BEGIN_SRC elisp
(use-package undo-tree
	:ensure t
	:commands(undo-tree-mode)
	:defer t
)
#+END_SRC
** Terminal-here
#+BEGIN_SRC emacs-lisp
(use-package terminal-here
	:ensure t
	:commands(terminal-here terminal-here-launch terminal-here-project-launch)
)
#+END_SRC
* Org Mode
#+BEGIN_SRC emacs-lisp
(use-package org
	:mode("\\.org\\'" . org-mode)
	:bind(:map org-mode-map
		("<f1>" . org-export-dispatch)
	)
	:custom
	(org-startup-with-inline-images nil)
	(org-latex-listings 'minted)
	(org-latex-packages-alist '(("" "minted")))
	(org-latex-minted-options
		'(
			("frame" "lines")
			("linenos" "")
		))
	(org-latex-pdf-process
	'("pdflatex -shell-escape -interaction=nonstopmode %f"
	"pdflatex -shell-escape -interaction=nonstopmode %f"
	))
	(org-latex-toc-command "\\tableofcontents \\clearpage")
)

;; Call load ox-latex only when exporting
(use-package ox-latex
	:no-require t
	:commands org-export-dispatch
)

(use-package helm-org-rifle
	:ensure t
	:bind(:map org-mode-map
	("M-s" . helm-org-rifle-org-directory)
)
)

(use-package ox-twbs
	:ensure t
	:defer t
)

(use-package org-wc
	:ensure t
	:commands(org-wc-display)
)

#+END_SRC
** Enable python snippets

#+BEGIN_SRC emacs-lisp
(org-babel-do-load-languages
 'org-babel-load-languages
 '((python . t)))
#+END_SRC

* Programming
** C# / .NET
#+BEGIN_SRC emacs-lisp
(use-package csharp-mode
	:mode ("\\.cs\\'" . csharp-mode)
	:ensure t
	:hook(csharp-mode . (lambda ()
		(dotnet-mode 1)
		(omnisharp-mode 1)
	))
	:config(lambda()
	(eval-after-load
	'company
	'(add-to-list 'company-backends 'company-omnisharp))
)
)

(use-package dotnet
	:ensure t
	:diminish
	:commands(dotnet-mode)
)

(use-package csproj-mode
	:mode("\\.csproj\\'" . csproj-mode)
	:ensure t
	:diminish
)

(use-package omnisharp
	:ensure t
	:after csharp-mode
	:commands(omnisharp-mode)
)

#+END_SRC
** Debugging =`DAP-mode=
#+BEGIN_SRC emacs-lisp
(use-package dap-mode
	:ensure  t
	:diminish
	:defer t
	:commands(dap-mode)
)
#+END_SRC
** Flycheck
#+BEGIN_SRC elisp
(use-package flycheck
	:ensure t
	:diminish
	:hook(c-mode-common . (lambda ()
	(flycheck-mode)
	(flycheck-select-checker 'c/c++-gcc)
))
	:commands(flycheck-mode)
)
#+END_SRC
** Company
#+BEGIN_SRC emacs-lisp
(use-package company
	:ensure t
	:diminish
	:hook((web-mode css-mode c-mode-common csharp-mode) . company-mode)
	:custom(comany-idle-delay 0.3)
)

(use-package company-web
	:ensure t
	:diminish
	:after company
)

(use-package company-lsp
  :defer t
  :after company lsp
  :custom (company-lsp-cache-candidates 'auto)

)

(use-package company-quickhelp
	:ensure t
	:diminish
	:after company
	:hook (company-mode . company-quickhelp-mode)
)

#+END_SRC

** LSP-Mode
LSP-mode helps with managing LSP's (Language Server Protocols)
#+BEGIN_SRC emacs-lisp
(use-package lsp-mode
	:ensure t
	:hook((c-mode-common) . lsp)
	:bind(:map c-mode-base-map
			("<f5>" . lsp-find-definition)
			("<f6>" . lsp-find-references)
			("<f7>" . lsp-find-declaration)

)
	:config(lsp-lens-mode t)
)

(use-package flycheck-clang-analyzer
  :ensure t
  :after flycheck
  :config (flycheck-clang-analyzer-setup))

(use-package lsp-ui
	:ensure t
	:after lsp-mode
	:bind(:map lsp-ui-mode-map
		("C-f" . lsp-ui-imenu)
		([remap xref-find-definitions] . lsp-ui-peek-find-definitions)
        ([remap xref-find-references] . lsp-ui-peek-find-references)
	)
	:custom
	(lsp-ui-doc-enable t)
	(lsp-ui-doc-header nil)
	(lsp-ui-doc-include-signature t)
	(lsp-ui-doc-position 'top)
	(lsp-ui-doc-border (face-foreground 'default))
	(lsp-ui-sideline-enable nil)
	(lsp-ui-sideline-ignore-duplicate t)
	(lsp-ui-sideline-show-code-actions nil)
	:diminish
	:commands lsp-ui-mode
)
#+END_SRC
** Latex
#+BEGIN_SRC emacs-lisp
(use-package latex
	:no-require t
	:hook(latex-mode . flyspell-mode)
)

(use-package latex-extra
	:ensure t
	:after latex
)
#+END_SRC
** Python Mode
#+BEGIN_SRC emacs-lisp
(use-package python
	:mode("\\.py\\'" . python-mode)
)

(use-package jedi
	:ensure t
	:commands(company-jedi)
)

(defun my/python-mode-hook ()
  (add-to-list 'company-backends 'company-jedi))

(add-hook 'python-mode-hook 'my/python-mode-hook)

(use-package blacken
	:ensure t
	:diminish
	:defer t
)

(use-package pylint
	:ensure t
	:diminish
	:defer t
)

(use-package virtualenvwrapper
	:ensure t
	:defer t
)

#+END_SRC
** C Mode
#+BEGIN_SRC emacs-lisp
(setq c-basic-offset 4)

(use-package function-args
	:ensure t
	:diminish
	:defer t
)

(defun luca/c-debug (directory)
	"Sets up debugging environment for c"
	(interactive "sFile Path:")
	(gdb (format "gdb -i=mi %s" directory ))
	(gbd-many-windows)
)



(use-package gud
	:no-require t
	:commands luca/c-debug
	:bind(:map c-mode-base-map
		("C-<f5>" . gud-break)
		("C-<f6>" . gud-step)
		("C-<f7>" . gud-next)
)
)

#+END_SRC
** HTML
#+BEGIN_SRC emacs-lisp
(use-package web-mode
	:ensure t
	:mode("\\.html\\'")
	:mode("\\.cshtml\\'")
	:config(web-mode-toggle-current-element-highlight)
	:bind(:map web-mode-map
	("C-c C-e -" . web-mode-element-contract)
	("C-c C-e +" . web-mode-element-extract)
	("C-c C-e /" . web-mode-element-close)
	("C-c C-e a" . web-mode-element-content-select)
	("C-c C-e i" . web-mode-element-insert)
	("C-c C-e w" . web-mode-element-wrap)
	("C-c C-e k" . web-mode-element-kill)
)
)

(use-package css-mode
	:ensure t
	:mode("\\.css\\'")
)

(use-package css-comb
	:ensure t
	:commands (css-comb)
)

(use-package simple-httpd
	:no-require t
	:after web-mode
	:ensure t
	:diminish
	:hook(web-mode . httpd-start)
)

(use-package impatient-mode
	:ensure t
	:after web-mode
	:hook((web-mode css-mode) . impatient-mode)
)

(use-package zencoding-mode
	:ensure t
	:diminish
	:hook(web-mode . zencoding-mode)
	:bind(:map web-mode-map("C-`" . zencoding-expand-line))
)

(use-package web-beautify
	:ensure t
	:diminish
	:defer t
)
#+END_SRC
** Yasnippet setup
To create snippets, just write it out on any buffer and leave a '~' right behind the placeholder. Then highlight the region and use the command =aya-create=. Use =aya-persist-snippet= to save the created snippet.
#+BEGIN_SRC emacs-lisp
(use-package yasnippet
	:ensure t
	:diminish
	:hook(prog-mode . yas-minor-mode)
	(org-mode . yas-minor-mode)
	:bind*("C-~" . yas-insert-snippet)
	:config
	(yas-reload-all)
	(use-package yasnippet-snippets
		:ensure t
		:after yasnippet
	)
	(use-package yasnippet-classic-snippets
		:ensure t
		:after yasnippet
	)
)

(use-package auto-yasnippet
	:ensure t
	:after yasnippet
	:commands(aya-create aya-expand aya-open-line)
	:diminish
)
#+END_SRC
** Smart Compile
#+BEGIN_SRC emacs-lisp
(use-package smart-compile
	:ensure t
	:diminish
	:commands(smart-compile)
)
#+END_SRC
** So-long
This helps mitigate the slowness which comes with opening large files in emacs.
#+BEGIN_SRC emacs-lisp
(use-package so-long
	:load-path("~/.emacs.d/packages/")
	:commands global-so-long-mode
	:init(global-so-long-mode)
)
#+END_SRC
* Key Bindings

config.el
(require 'package)
(setq package-enable-at-startup nil)

;;; remove SC if you are not using sunrise commander and org if you like outdated packages
(setq package-archives '(("ELPA"  . "http://tromey.com/elpa/")
			 ("gnu"   . "http://elpa.gnu.org/packages/")
			 ("melpa" . "https://melpa.org/packages/")
			 ("elpa"  . "https://elpa.gnu.org/packages/")
			 ))
(package-initialize)

(defconst emacs-start-time (current-time))

(defvar file-name-handler-alist-old file-name-handler-alist)

(setq package-enable-at-startup nil
      file-name-handler-alist nil
      message-log-max 16384
      gc-cons-threshold 402653184
      gc-cons-percentage 0.6
      auto-window-vscroll nil)

(add-hook 'after-init-hook
          `(lambda ()
             (setq file-name-handler-alist file-name-handler-alist-old
                   gc-cons-threshold 800000
                   gc-cons-percentage 0.1)
             (garbage-collect)) t)

(setq initial-major-mode (quote fundamental-mode))

(use-package benchmark-init
  :ensure t
  :init(benchmark-init/activate))

(delete-selection-mode 1)

(setq package-enable-at-startup nil)

(setq site-run-file nil)

(use-package doom-themes
	:diminish
	:ensure t
	:config (load-theme 'doom-city-lights t)
)

(setq inhibit-splash-screen t
initial-buffer-choice  nil
)

(add-hook 'prog-mode-hook 'display-line-numbers-mode)

(menu-bar-mode -1)
(toggle-scroll-bar -1)
(tool-bar-mode -1)

(setq ring-bell-function 'ignore)

(setq display-time-24hr-format t)
(setq display-time-format "%d %B %Y - %H:%M")
(display-time-mode 1)

(use-package dashboard
:ensure t
:config(setq dashboard-startup-banner "~/.emacs.d/images/KEC_Dark_BK.png")
(setq dashboard-banner-logo-title "Don't do the rain dance if you can't handle the thunder - Ken M")
(setq dashboard-items '((recents  . 3)
                        (bookmarks . 3)
                        (projects . 5)
                        (agenda . 0)
                        (registers . 0)))
(dashboard-setup-startup-hook)
)

(use-package beacon
	:ensure t
	:config
		(beacon-mode 1)
)

(setq-default
frame-title-format '("Lucinda?"))

(set-frame-font "Consolas 14" nil t)

(use-package treemacs
	:ensure t
	:diminish
	:bind*("C-<tab>" . treemacs)
	:custom
	(treemacs-indentation 2)
	(treemacs-eldoc-display t)
	(treemacs-show-hidden-files t)
	(treemacs-width 35)
	(treemacs-display-in-side-window nil)
	(treemacs-pre-file-insert-predicates)
)

(use-package treemacs-projectile
	:ensure t
	:diminish
	:after treemacs projectile
)

(use-package lsp-treemacs
	:ensure t
	:diminish
	:after treemacs lsp
	:config(lsp-treemacs-sync-mode 1)
)

(use-package awesome-pair
	:load-path("~/.emacs.d/elpa/")
	:hook( (org-mode prog-mode) . awesome-pair-mode)
	:bind
	("(" . awesome-pair-open-round)
	("[" . awesome-pair-open-bracket)
	("{" . awesome-pair-open-curly)
	("=" . awesome-pair-equal)
	("M-\"" . awesome-pair-wrap-double-quote)
	("M-[" . awesome-pair-wrap-bracket)
	("M-{" . awesome-pair-wrap-curly)
	("M-(" . awesome-pair-wrap-round)
	("M-)" . awesome-pair-unwrap)
)

(use-package projectile
	:diminish
	:ensure t
	:init(projectile-mode)
	(require 'cc-mode)
	:bind("C-x r p" . projectile-switch-project)
		 ("C-x C-M-f" . projectile-find-file)
		 ("C-x C-M-s" . projectile-save-project-buffers)
	:bind(:map c-mode-base-map
		("<f1>" . projectile-compile-project))
		("<f2>" . luca/c-debug)
)

(use-package helm-projectile
	:ensure t
	:after projectile
	:bind("C-M-s" . helm-projectile-ag)
	:init(helm-projectile-on)
)

(use-package magit
	:ensure t
	:diminish
	:commands(magit-stage-file magit-status magit-commit-create)
	:bind(:map prog-mode-map
		("C-c t" . magit-stage-file)
		("C-c s" . magit-status)
		("C-c c" . magit-commit-create)
	)
)

(use-package vc
	:no-require t
	:defer t
	:commands(vc diff)
	:bind("C-c d" . vc-diff)
)

(use-package git-gutter
	:ensure t
	:diminish
	:hook (prog-mode . git-gutter-mode)
	:hook (magit-post-refresh . git-gutter:update-all-windows)
	:bind(:map prog-mode-map
		("C-c n" . git-gutter:next-hunk)
 		("C-c p" . git-gutter:previous-hunk)
)
)

(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
;; backwards compatibility as default-buffer-file-coding-system
;; is deprecated in 23.2.
(if (boundp 'buffer-file-coding-system)
	(setq-default buffer-file-coding-system 'utf-8)
	(setq default-buffer-file-coding-system 'utf-8))

;; Treat clipboard input as UTF-8 string first; compound text next, etc.
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

(setq make-backup-files nil)
(setq auto-save-default nil)

(defalias 'yes-or-no-p 'y-or-n-p)

(global-visual-line-mode)

(use-package which-key
	:ensure t
	:diminish
	:config (which-key-mode)
	)

(use-package ispell
	:no-require t
	:defer t
	:bind (:map org-mode-map("C-<return>" . ispell-word))
)

(use-package define-word
	:diminish
	:ensure t
	:defer t
)

(use-package helm-ag
	:ensure t
	:diminish
	:commands(helm-ag)
)

(global-unset-key "\C-z")
(global-unset-key "\C-x\C-z")
(global-unset-key "\C-x\C-c")

(use-package fill
	:no-require t
	:bind("<f9>" . fill-region )
)

(use-package terminal-here
	:ensure t
	:commands(terminal-here terminal-here-launch terminal-here-project-launch)
)

(use-package org
	:mode("\\.org\\'" . org-mode)
	:bind(:map org-mode-map
		("<f1>" . org-export-dispatch)
	)
	:custom
	(org-startup-with-inline-images nil)
	(org-latex-listings 'minted)
	(org-latex-packages-alist '(("" "minted")))
	(org-latex-minted-options
		'(
			("frame" "lines")
			("linenos" "")
		))
	(org-latex-pdf-process
	'("pdflatex -shell-escape -interaction=nonstopmode %f"
	"pdflatex -shell-escape -interaction=nonstopmode %f"
	))
	(org-latex-toc-command "\\tableofcontents \\clearpage")
)

;; Call load ox-latex only when exporting
(use-package ox-latex
	:no-require t
	:commands org-export-dispatch
)

(use-package helm-org-rifle
	:ensure t
	:bind(:map org-mode-map
	("M-s" . helm-org-rifle-org-directory)
)
)

(use-package ox-twbs
	:ensure t
	:defer t
)

(use-package org-wc
	:ensure t
	:commands(org-wc-display)
)

(org-babel-do-load-languages
 'org-babel-load-languages
 '((python . t)))

(use-package csharp-mode
	:mode ("\\.cs\\'" . csharp-mode)
	:ensure t
	:hook(csharp-mode . (lambda ()
		(dotnet-mode 1)
		(omnisharp-mode 1)
	))
	:config(lambda()
	(eval-after-load
	'company
	'(add-to-list 'company-backends 'company-omnisharp))
)
)

(use-package dotnet
	:ensure t
	:diminish
	:commands(dotnet-mode)
)

(use-package csproj-mode
	:mode("\\.csproj\\'" . csproj-mode)
	:ensure t
	:diminish
)

(use-package omnisharp
	:ensure t
	:after csharp-mode
	:commands(omnisharp-mode)
)

(use-package dap-mode
	:ensure  t
	:diminish
	:defer t
	:commands(dap-mode)
)

(use-package company
	:ensure t
	:diminish
	:hook((web-mode css-mode c-mode-common csharp-mode) . company-mode)
	:custom(comany-idle-delay 0.3)
)

(use-package company-web
	:ensure t
	:diminish
	:after company
)

(use-package company-lsp
  :defer t
  :after lsp
  :custom (company-lsp-cache-candidates 'auto)

)

(use-package company-quickhelp
	:ensure t
	:diminish
	:hook (company-mode . company-quickhelp-mode)
)

(use-package lsp-mode
	:ensure t
	:hook((c-mode-common) . lsp)
	:bind(:map c-mode-base-map
			("<f5>" . lsp-find-definition)
			("<f6>" . lsp-find-references)
			("<f7>" . lsp-find-declaration)

)
	:config(lsp-lens-mode t)
)

(use-package flycheck-clang-analyzer
  :ensure t
  :after flycheck
  :config (flycheck-clang-analyzer-setup))

(use-package lsp-ui
	:ensure t
	:after lsp-mode
	:bind(:map lsp-ui-mode-map
		("C-f" . lsp-ui-imenu)
		([remap xref-find-definitions] . lsp-ui-peek-find-definitions)
        ([remap xref-find-references] . lsp-ui-peek-find-references)
	)
	:custom
	(lsp-ui-doc-enable t)
	(lsp-ui-doc-header nil)
	(lsp-ui-doc-include-signature t)
	(lsp-ui-doc-position 'top)
	(lsp-ui-doc-border (face-foreground 'default))
	(lsp-ui-sideline-enable nil)
	(lsp-ui-sideline-ignore-duplicate t)
	(lsp-ui-sideline-show-code-actions nil)
	:diminish
	:commands lsp-ui-mode
)

(use-package latex
	:no-require t
	:hook(latex-mode . flyspell-mode)
)

(use-package latex-extra
	:ensure t
	:after latex
)

(use-package python
	:mode("\\.py\\'" . python-mode)
)

(use-package jedi
	:ensure t
	:commands(company-jedi)
)

(defun my/python-mode-hook ()
  (add-to-list 'company-backends 'company-jedi))

(add-hook 'python-mode-hook 'my/python-mode-hook)

(use-package blacken
	:ensure t
	:diminish
	:defer t
)

(use-package pylint
	:ensure t
	:diminish
	:defer t
)

(use-package virtualenvwrapper
	:ensure t
	:defer t
)

(setq c-basic-offset 4)

(use-package function-args
	:ensure t
	:diminish
	:defer t
)

(defun luca/c-debug (directory)
	"Sets up debugging environment for c"
	(interactive "sFile Path:")
	(gdb (format "gdb -i=mi %s" directory ))
	(gbd-many-windows)
)



(use-package gud
	:no-require t
	:commands luca/c-debug
	:bind(:map c-mode-base-map
		("C-<f5>" . gud-break)
		("C-<f6>" . gud-step)
		("C-<f7>" . gud-next)
)
)

(use-package web-mode
	:ensure t
	:mode("\\.html\\'")
	:mode("\\.cshtml\\'")
	:config(web-mode-toggle-current-element-highlight)
	:bind(:map web-mode-map
	("C-c C-e -" . web-mode-element-contract)
	("C-c C-e +" . web-mode-element-extract)
	("C-c C-e /" . web-mode-element-close)
	("C-c C-e a" . web-mode-element-content-select)
	("C-c C-e i" . web-mode-element-insert)
	("C-c C-e w" . web-mode-element-wrap)
	("C-c C-e k" . web-mode-element-kill)
)
)

(use-package css-mode
	:ensure t
	:mode("\\.css\\'")
)

(use-package css-comb
	:ensure t
	:commands (css-comb)
)

(use-package simple-httpd
	:no-require t
	:after web-mode
	:ensure t
	:diminish
	:hook(web-mode . httpd-start)
)

(use-package impatient-mode
	:ensure t
	:after web-mode
	:hook((web-mode css-mode) . impatient-mode)
)

(use-package zencoding-mode
	:ensure t
	:diminish
	:hook(web-mode . zencoding-mode)
	:bind(:map web-mode-map("C-`" . zencoding-expand-line))
)

(use-package web-beautify
	:ensure t
	:diminish
	:defer t
)

(use-package yasnippet
	:ensure t
	:diminish
	:hook(prog-mode . yas-minor-mode)
	(org-mode . yas-minor-mode)
	:bind*("C-~" . yas-insert-snippet)
	:config
	(yas-reload-all)
	(use-package yasnippet-snippets
		:ensure t
		:after yasnippet
	)
)

(use-package auto-yasnippet
	:ensure t
	:after yasnippet
	:commands(aya-create aya-expand aya-open-line)
	:diminish
)

(use-package smart-compile
	:ensure t
	:diminish
	:commands(smart-compile)
)

(use-package so-long
	:load-path("~/.emacs.d/packages/")
	:commands global-so-long-mode
	:init(global-so-long-mode)
)

(use-package ace-window
	:ensure t
	:diminish
	:commands ace-window
	:bind*("M-o" . ace-window)
)

(use-package multiple-cursors
	:ensure t
	:commands (mc/mark-next-like-this mc/mark-previous-like-this)
	:diminish
	:bind
	("C->" . mc/mark-next-like-this)
	("C-<" . mc/mark-previous-like-this)
)

(use-package ace-mc
	:ensure t
	:commands (ace-mc-add-multiple-cursors)
	:diminish
	:bind
	("C-M->" . ace-mc-add-multiple-cursors)
	("C-M-<" . ace-mc-add-multiple-cursors)
)

(global-set-key (kbd "S-C-<left>") 'shrink-window-horizontally)
(global-set-key (kbd "S-C-<right>") 'enlarge-window-horizontally)
(global-set-key (kbd "S-C-<down>") 'shrink-window)
(global-set-key (kbd "S-C-<up>") 'enlarge-window)

(use-package helm
	:ensure t
	:diminish
	:bind
	("C-s" . helm-occur)
	("M-y" . helm-show-kill-ring)
	("C-x r m" . helm-bookmarks)
	("C-x C-b" . helm-buffers-list)
	("C-x C-f" . helm-find-files)
)

(use-package ace-jump-mode
	:ensure t
	:bind
	("C-z" . ace-jump-mode)
	("M-z" . ace-jump-mode-pop-mark)
)

(global-set-key (kbd "C-M-g") 'query-replace-regexp)
(use-package smex
	:ensure t
	:diminish
	:bind("M-x" . smex)
)

(use-package simple
	:no-require t
	:hook(before-save . delete-trailing-whitespace)
)

(use-package isearch
	:no-require t
	:bind
	("M-s" . isearch-forward)
	("M-r" . isearch-backward)
)

(use-package senator
	:no-require t
	:bind*("<M-down>" . senator-transpose-tags-down)
	("<M-up>" . senator-transpose-tags-up)
)

(use-package tramp :defer t)
(use-package with-editor :defer t)
(use-package org-agenda :defer t)
(use-package speedbar :defer t)
(use-package gud :defer t)
(use-package smartscan :defer t)
(use-package ivy :diminish :defer t)

(setq gc-cons-threshold 16777216
      gc-cons-percentage 0.1)
